python/ray/ray_logging.py

```diff
@@ -0,0 +1,59 @@
+import logging
+import os
+
+_default_handler = None
+
+
+def setup_logger(logging_level, logging_format):
+    """Setup default logging for ray."""
+    logger = logging.getLogger("ray")
+    if type(logging_level) is str:
+        logging_level = logging.getLevelName(logging_level.upper())
+    logger.setLevel(logging_level)
+    global _default_handler
+    if _default_handler is None:
+        _default_handler = logging.StreamHandler()
+        logger.addHandler(_default_handler)
+    _default_handler.setFormatter(logging.Formatter(logging_format))
+    logger.propagate = False
+
+
+class StandardStreamInterceptor:
+    """Used to intercept stdout and stderr.
+
+    Intercepted messages are handled by the given logger.
+
+    NOTE: The logger passed to this method should always have
+          logging.INFO severity level.
+
+    Example:
+        >>> from contextlib import redirect_stdout
+        >>> logger = logging.getLogger("ray_logger")
+        >>> hook = StandardStreamHook(logger)
+        >>> with redirect_stdout(hook):
+        >>>     print("a") # stdout will be delegated to logger.
+
+    Args:
+        logger: Python logger that will receive messages streamed to
+                the standard out/err and delegate writes.
+        intercept_stdout(bool): True if the class intercepts stdout. False
+                         if stderr is intercepted.
+    """
+
+    def __init__(self, logger, intercept_stdout=True):
+        self.logger = logger
+        self.intercept_stdout = intercept_stdout
+
+    def write(self, message):
+        """Redirect the original message to the logger."""
+        self.logger.info("{}".format(message))
+
+    def flush(self):
+        # flush method can be empty because logger will handle flush
+        for handler in self.logger.handlers:
+            handler.flush()
+
+    def isatty(self):
+        # Return the standard out isatty. This is used by colorful.
```

Comment:
Btw, this is one drawback of this approach. If there are libraries that use some file-like interface (ex, fileno) with stdout it can cause issues.

Reply:
If there are users who use interesting libraries to challenge stdout and stderr, we should probably need to come up with hacky solution to obtain file handle within the logger.

Reply:
Doesn't seem like a big deal